include/envoy/event/dispatcher.h

```diff
@@ -215,6 +217,13 @@ class Dispatcher {
    * current thread of execution is on the same thread upon which the dispatcher loop is running.
    */
   virtual bool isThreadSafe() const PURE;
+
+  /**
+   * Registers callback to be called in the event loop prior to polling for
+   * events. Overwrites |callback| from previous calls. Calling with null
+   * callback unregisters previously registered callback.
+   */
+  virtual void registerOnPrepareCallback(OnPrepareCallback callback) PURE;
```

Comment:
From an interface perspective I don't think this makes sense. At the dispatcher level if we are going to do this we should allow a list of callbacks, otherwise the callers will not understand who actually owns this callback.

Stepping back, is any of this change really necessary? libevent already actually keeps track of an approximate time. Could we instead expose libevent's approximate time as an API and use that? Stepping back even further, is this a premature perf optimization or do we know this is an issue?

Reply:
> How does the ownership has to do with a list v.s. a individual variable?

Because from an interface perspective it's not clear who currently owns the prepare callback. The code has to be searched. I don't think that makes sense, so if we want to add this type of interface we should be able to add a callback and remove a callback and the internal implementation is a list.

In general though I wonder if it would be better to just build an approximateNow() API into the dispatcher itself, which can internally use the prepare callback, etc.

Reply:
That's also an option.  In fact we had that in an earlier stage on this PR, see https://github.com/envoyproxy/envoy/pull/8494/files/188b1078eaf4a49300f6a604a6cac60071267faa.  Is that something you had in mind?

Reply:
Yeah if I had to choose I would prefer ^ but if you want to do a general API and make it an add/remove pair with an internal list that's fine too.

Reply:
> > How does the ownership has to do with a list v.s. a individual variable?
> 
> Because from an interface perspective it's not clear who currently owns the prepare callback. The code has to be searched. I don't think that makes sense, so if we want to add this type of interface we should be able to add a callback and remove a callback and the internal implementation is a list.

I don't have strong preference on Dispatcher interface. Can we change the new interface to take `OnPrepareCallback&&` so it's clear that Dispatcher owns the callback? I guess more than the ownership the callback life time is a bit confusing. Do you think instead of using std::bind(), adding a callback interface to event scheduler is clearer?  Either Dispatcher or EnvoyQuicClock can implement that interface and call registerCallback() and unregisterCallback() explicitly before itself is destroyed.

> In general though I wonder if it would be better to just build an approximateNow() API into the dispatcher itself, which can internally use the prepare callback, etc.

We took this approach before, but I think we need to latch the value in EnvoyQuicClock before each event loop anyway so that EnvoyQuicClock::Now() can update the timestamp as well. If you think  it worth to keep the approximate now in Dispatcher, we can add such field in it too.

Reply:
re: the callback interface, if we keep it I care mostly that the interface supports an add/remove API, even if the internal implementation asserts right now that there is a single callback (or just use a list). 

> We took this approach before, but I think we need to latch the value in EnvoyQuicClock before each event loop anyway so that EnvoyQuicClock::Now() can update the timestamp as well.

Couldn't EnvoyQuicClock::Now() just call the dispatcher's approximate now API?

Reply:
> Couldn't EnvoyQuicClock::Now() just call the dispatcher's approximate now API?

For congestion control, we want the time as accurate as possible.

Reply:
I don't follow. How is using the same value that would be snapped in the dispatcher's internal prepare callback any more/less accurate then you copying it in your own prepare callback?

Reply:
> I don't follow. How is using the same value that would be snapped in the dispatcher's internal prepare callback any more/less accurate then you copying it in your own prepare callback?

I was referring to using dispatcher Approximate now for QUIC Now() is not accurate. But as Bence's last commit, Now() can update dispatcher Approximate now before using it. I'm good with this approach.